### ISSUE TITLE
2D BSL files converted into h5 format no longer loading in SasView 4.2.2

### ISSUE THREAD

**Issue by smk78** [SasView/sasview#1386]
In testing issue #1383 it was discovered that 2D BSL files converted into h5 format were no longer loading into SasView 4.2.2, SasView running from master, or SasView running from the converting_bslloader_python_4x_branch. They _do_ however load in SasView 4.1.2.

More curiously, 2D ASCII files converted into h5 format load fine in all these versions.

**Comment by smk78** [comment]
[issue-1386.zip](https://github.com/SasView/sasview/files/3651671/issue-1386.zip)

**Comment by smk78** [comment]
Uploaded issue-1386.zip contains 7 files:

- X2500n.L2D , the original 2D BSL data (n=0: header, n=1: SANS, n=2: calibration (here blank), n=3: WANS)
- x25000-pr.h5 , converted file output by converting_bslloader_python_4x branch
- x25000-422.h5 , converted file output by SasView 4.2.2
- x25000-421.h5 , converted file output by SasView 4.2.1

**Comment by smk78** [comment]
The above h5 files all load and display in SasView 4.1.2

The above files _do not_ load in SasView 4.2.2. The Console reports:

```
The Data file you selected could not be loaded.
Make sure the content of your file is properly formatted.
When contacting the SasView team, mention the following:
too many indices for arrayThe following issues were found whilst loading x25000-pr.h5:
```

The above files _do not_ load in SasView 4.x master.  The Command Window reports:

```
10:39:41 - INFO : sas.sascalc.dataloader.readers.xml_reader:  78: Start tag expected, '<' not found, line 1, column 1
10:39:41 - INFO : sas.sascalc.dataloader.readers.xml_reader:  78: Start tag expected, '<' not found, line 1, column 1
```

which is curious as they are h5 files...

**Comment by smk78** [comment]
There are also differences in the size of the files (even though they were converted from the same input file with the same options and metadata entries):

- x25000-pr.h5 , 169 KB
- x25000-422.h5 , 130 KB
- x25000-421.h5 , 215 KB

**Comment by smk78** [comment]
@krzywon is this something you could take a look at?

**Comment by krzywon** [comment]
I have a fix for this issue in branch ticket-1386 and will submit a PR after checking a few more things, but I am more concerned about the data the BSL converter is producing. The data is in pixel space, not Q space. The NXcanSAS standard 
> is an implementation of the canSAS standard for storing both one-dimensional and multi-dimensional _reduced_ small-angle scattering data. http://download.nexusformat.org/doc/html/classes/applications/NXcanSAS.html

We removed a reader because the data was in pixel space. Should we modify the writer to output in Q space instead?

**Comment by smk78** [comment]
Interesting...

I have two observations:

1. The x25000 2D BSL example I provided above has no Q-axis (.QAX) files. I don't know why. Sorry. So I'm attaching a new example here which does have them.
[LOQ_107487_QXY.zip](https://github.com/SasView/sasview/files/3690710/LOQ_107487_QXY.zip)
I could understand the File Converter producing pixel space data if the Q-axis information was missing (though note, this is image pixel space not raw detector pixel space).

2. However, when I cam to try the new example I noticed that when you select 2D BSL in the File Converter, it greys out the Q-Axis Data input box, so you can't currently provide the Q-axis data anyway!

**Comment by butlerpd** [comment]
I believe in principle that this ticket should have been closed when the PR request above was approved and merged. @smk78 can you concur please? If so we can close this issue

**Comment by smk78** [comment]
No, leave this open. There is still a problem with 4.x. Use the File Converter in 5.0.2 to convert the \test\convertible_files X25000.L2D or the LOQ_107487_QXY.TXT above. Both are converted to .h5, read back in to 5.0.2 and display correctly. If you then try and load the .h5 into 4.2.2 _only_ the 107487 will be loaded and display. The X25000 example generates this error:

> 2020-04-27 12:41:06 : The following issues were found whilst loading 25000.h5:
The Data file you selected could not be loaded.
Make sure the content of your file is properly formatted.
When contacting the SasView team, mention the following:
too many indices for array

**Comment by krzywon** [comment]
@smk78 The changes I made for this ticket have not been incorporated into a 4.x release yet, but are in master. I just converted X25000.L2D using 5.0.2 and loaded into SasView run from the master branch. Are you having issues loading there as well?

**Comment by smk78** [comment]
You are right, @krzywon, 4.x master loads the converted X25000 h5 file.

**Comment by smk78** [comment]
So guess this can be closed after all?

**Comment by butlerpd** [comment]
Yes ... if it is fixed in master it should be closed (as a ticket fixed in 4.3 release :-) Closing now then.
